File: CorrectionsCSharp/Chapitre 4/1_Structures_classes/1X01_DATES_CORRECTION/Application.cs
using System;

public class Application
{
    public void fonctionPrincipale()
    {
        //DEBUT de votre programme
		Date d = new Date();
		d.afficher();

        String prenom, nom;
        Console.WriteLine("Prenom : ");
        prenom = Utilisateur.saisirTexte();
		if (prenom.StartsWith("P") == true) {
			string prenomMajuscules = prenom.ToUpper();
			Console.WriteLine(prenomMajuscules);
		}
		

        Console.WriteLine("Nom :");
        nom = Utilisateur.saisirTexte();

        Console.WriteLine("Entrez votre date de naissance");
        Date dateNaissance = creerDate();
        Date dateEmbauche = null;
        do
        {
            Console.WriteLine("Entrez votre date d'embauche");
            dateEmbauche = creerDate();
        } while (estPlusRecenteQue(dateNaissance, dateEmbauche) == true);


        Console.WriteLine("Bonjour {0} {1}, voici votre date de naissance :", prenom, nom);
        dateNaissance.afficher();
		
        Console.WriteLine("et voici votre date d'embauche : ");
        dateEmbauche.afficher();

		

        Utilisateur.saisirTexte();
        //FIN de votre programme
    }

    Date creerDate()
    {
        Date nvDate = new Date();
        do
        {
            Console.Write("Année : ");
            nvDate.annee = Utilisateur.saisirEntier();
        } while (nvDate.annee <= 0);

        do
        {
            Console.Write("Mois : ");
            nvDate.mois = Utilisateur.saisirEntier();
        } while (nvDate.mois < 1 || nvDate.mois > 12);

        int jourMax = obtenirJourMax(nvDate.mois, nvDate.annee);
        do
        {
            Console.Write("Jour : ");
            nvDate.jour = Utilisateur.saisirEntier();
        } while (nvDate.jour < 1 || nvDate.jour > jourMax);

        return nvDate;
    }

	int obtenirJourMax(int mois, int annee) {
		int jourMax = 31;
		if (mois == 4 
			|| mois == 6 
			|| mois == 9 
			|| mois == 11) {
			jourMax = 30;
		} else if (mois == 2) {
			if (estBissextile(annee) == true) {
				jourMax = 29;
			} else {
				jourMax = 28;
			}
		}
		return jourMax;
	}

    bool estBissextile(int annee)
    {
        bool estBissextile = false;
		if (annee % 4 == 0 
			&& (annee % 100 != 0 || annee % 400 == 0)) {
			estBissextile = true;
		}

		return estBissextile;
    }

    

    bool estPlusRecenteQue(Date date1, Date date2)
    {
        bool estPlusRecente = true;

        if (date1.annee < date2.annee)
        {
            estPlusRecente = false;
        }
        else if (date1.annee == date2.annee && date1.mois < date2.mois)
        {
            estPlusRecente = false;
        }
        else if (date1.annee == date2.annee && date1.mois == date2.mois && date1.jour <= date2.jour)
        {
            estPlusRecente = false;
        }

        return estPlusRecente;
    }

    //DECLAREZ VOS FONCTIONS EN DESSOUS DE CETTE LIGNE

    /* EXEMPLES DE PROCEDURES ET FONCTIONS :
	
		void uneProcedure(int param1, int param2)
		{
			Console.WriteLine(param1+param2);
		{
		
		int uneFonction(int param1, int param2)
		{
			int resultat = 0;
			resultat = param1 + param2;
			return resultat;
		}
		
		
		*/

    //NE PAS TOUCHER LE CODE EN DESSOUS DE CETTE LIGNE
}

File: Java/Corrections/Chapitre 4/Section2_Dates/src/fr/purplegiraffe/Application.java
package fr.purplegiraffe;

import fr.purplegiraffe.NE_PAS_TOUCHER.ConsoleReader;

public class Application {
    public void fonctionPrincipale() {
        String prenom = ConsoleReader.readString("Prénom");
        String nom = ConsoleReader.readString("Nom");
        System.out.println("Date de naissance");
        Date naissance = creerDate();
        Date embauche;
        boolean naissancePlusRecente;
        do {
            System.out.println("Date d'embauche");
            embauche = creerDate();
            naissancePlusRecente = estPlusRecenteQue(naissance,embauche);
            if (naissancePlusRecente) {
                System.out.println("Vous etes né après votre embauche ?");
            }
        } while(naissancePlusRecente == true);

        System.out.format("Bonjour %s %s voici votre date de naissance : \n",prenom, nom);
        afficherDate(naissance);
        System.out.println("Et voici votre date d'embauche");
        afficherDate(embauche);
    }

    void afficherDate(Date d) {
        System.out.format("%02d/%02d/%04d\n",d.jour, d.mois, d.annee);
    }

    Date creerDate() {
        Date nvDate = new Date();

        nvDate.annee = ConsoleReader.readInt("année");
        do {
            nvDate.mois = ConsoleReader.readInt("mois");
        } while (nvDate.mois < 1 || nvDate.mois > 12);

        int jourMax = obtenirJourMax(nvDate);
        do {
            nvDate.jour = ConsoleReader.readInt("jour");
        } while (nvDate.jour < 1 || nvDate.jour > jourMax);

        return nvDate;
    }

    int obtenirJourMax(Date nvDate) {
        int jourMax;
        switch (nvDate.mois) {
            case 4:
            case 6:
            case 9:
            case 11:
                jourMax = 30;
                break;
            case 2:
                if (estBissextile(nvDate.annee) == true) {
                    jourMax = 29;
                } else {
                    jourMax = 28;
                }
                break;
            default:
                jourMax = 31;
        }
        return jourMax;
    }

    boolean estBissextile(int annee) {
        return (annee % 4 == 0 && annee % 100 != 0) || annee % 400 == 0;
    }

    boolean estPlusRecenteQue(Date d1, Date d2) {
        boolean d1PlusRecente = true;

        if (d2.annee > d1.annee) {
            d1PlusRecente = false;
        } else if (d1.annee == d2.annee) {
            if (d2.mois > d1.mois) {
                d1PlusRecente = false;
            } else if (d1.mois == d2.mois && d2.jour > d1.jour) {
                d1PlusRecente = false;
            }
        }
        return d1PlusRecente;
    }
}
